packages/arb-token-bridge-ui/src/util/withdrawals/fetchTokenWithdrawalsFromEventLogs.ts

```diff
@@ -5,42 +5,71 @@ import { Erc20Bridger } from '@arbitrum/sdk'
  * Fetches initiated token withdrawals from event logs in range of [fromBlock, toBlock].
  *
  * @param query Query params
- * @param query.fromAddress Address that initiated the withdrawal
- * @param query.toAddress Address that will receive the funds
+ * @param query.address Address that initiated the withdrawal, or received the funds
  * @param query.fromBlock Start at this block number (including)
  * @param query.toBlock Stop at this block number (including)
  * @param query.l2Provider Provider for the L2 network
  * @param query.l2GatewayAddresses L2 gateway addresses to use
  */
 export async function fetchTokenWithdrawalsFromEventLogs({
-  fromAddress,
-  toAddress,
+  sender,
+  receiver,
   fromBlock,
   toBlock,
   l2Provider,
   l2GatewayAddresses = []
 }: {
-  fromAddress?: string
-  toAddress?: string
+  sender?: string
+  receiver?: string
   fromBlock: BlockTag
   toBlock: BlockTag
   l2Provider: Provider
   l2GatewayAddresses?: string[]
 }) {
   const erc20Bridger = await Erc20Bridger.fromProvider(l2Provider)
 
-  return (
-    await Promise.all(
-      l2GatewayAddresses.map(gatewayAddress =>
+  const promises = l2GatewayAddresses.flatMap(gatewayAddress => {
+    const events = []
+
+    // funds sent by this address
+    if (sender) {
+      events.push(
         erc20Bridger.getL2WithdrawalEvents(
           l2Provider,
           gatewayAddress,
           { fromBlock, toBlock },
           undefined,
-          fromAddress,
-          toAddress
+          sender,
+          undefined
         )
       )
-    )
-  ).flat()
+    }
+
+    // funds received by this address
+    if (receiver) {
```

Comment:
When `sender` and `receiver` are the same values, and not undefined, `erc20Bridger.getL2WithdrawalEvents (RPC calls++)` and events-push function would run twice for the event. Is that expected, and can we optimise this?

Reply:
That's different event logs tho, one is fetching transactions sent by the address, the other is fetching transactions received by the address. So we need both even if sent and received values are defined and of the same value.